app/lua/lflash.c

```diff
@@ -0,0 +1,224 @@
+    G(L)->ROstrt.hash = cast(GCObject **, fh->pROhash);
+    G(L)->ROstrt.nuse = fh->nROuse ;
+    G(L)->ROstrt.size = fh->nROsize;
+    G(L)->ROpvmain    = cast(Proto *,fh->mainProto);
+  }
+}
+
+#define BYTE_OFFSET(t,f) cast(size_t, &(cast(t *, NULL)->f))
+/*
+ * Rehook address chain to correct Flash byte addressed within the mapped adress space 
+ * Note that on input each 32-bit address field is split into 2×16-bit subfields
+ *  -  the lu_int16 offset of the target address being referenced
+ *  -  the lu_int16 offset of the next address pointer. 
+ */
+
+static int rebuild_core (int fd, uint32_t size, lu_int32 *buf) {
+  int bi;  /* byte offset into memory mapped LFS of current buffer */ 
+  int wNextOffset = BYTE_OFFSET(FlashHeader,mainProto)/sizeof(lu_int32);
+  int wj;  /* word offset into current input buffer */
+  for (bi = 0; bi < size; bi += FLASH_PAGE_SIZE) {
+    int wi   = bi / sizeof(lu_int32);
+    int blen = ((bi + FLASH_PAGE_SIZE) < size) ? FLASH_PAGE_SIZE : size - bi;
+    int wlen = blen / sizeof(lu_int32);
+    if (vfs_read(fd, buf , blen) != blen)
+      return 0;
+
+    for (wj = 0; wj < wlen; wj++) {
+      if ((wi + wj) == wNextOffset) {  /* this word is the next linked address */
+        int wTargetOffset = buf[wj]&0xFFFF;
+        wNextOffset = buf[wj]>>16;
+        lua_assert(!wNextOffset || (wNextOffset>(wi+wj) && wNextOffset<size/sizeof(lu_int32)));
+        buf[wj] = cast(lu_int32, flashAddr + wTargetOffset*sizeof(lu_int32));
+      }
+    }
+    flashBlock(buf, blen);
+  }
+  return size;
+}
+
+
+/*
+ * Library function called by node.flash.load(filename).
+ */
+LUALIB_API int luaN_reload_reboot (lua_State *L) {
+  int fd, status;
+  FlashHeader fh;
+
+  const char *fn = lua_tostring(L, 1);
+  if (!fn || !(fd = vfs_open(fn, "r")))
+    return 0;
+   
+  if (vfs_read(fd, &fh, sizeof(fh)) != sizeof(fh) ||
+      fh.flash_sig != FLASH_SIG)
+    return 0;
+
+  if (vfs_lseek(fd, -1, VFS_SEEK_END) != fh.flash_size-1 ||
+      vfs_lseek(fd, 0, VFS_SEEK_SET) != 0)
+    return 0;
+
+  lu_int32 *buffer = luaM_newvector(L, FLASH_PAGE_SIZE / sizeof(lu_int32), lu_int32);
+
+  /*
+   * This is the point of no return.  We attempt to rebuild the flash.  If there
+   * are any problems them the Flash is going to be corrupt, so the only fallback
+   * is to erase it and reboot with a clean but blank flash.  Otherwise the reboot
+   * will load the new LFS.
+   *
+   * Note that the Lua state is not passed into the lua core because from this 
+   * point on, we make no calls on the Lua RTS.
+   */
+  flashErase(0,-1); 
+  if (rebuild_core(fd, fh.flash_size, buffer) != fh.flash_size)
+    flashErase(0,-1);  
+  /*
+   * Forcing a H/W timeout is the only robust way to insure that other interrupts
+   * and callbacks don't fire and attempt to reference to old LFS context.
+   */
+  while (1) {}
+
+  return 0;
+}
+
+
+/*
+ * Return a C closure pointing to the Flash Index function
+ */
+LUAI_FUNC int luaN_index (lua_State *L) {
```

Comment:
Currently one seems to need to do `node.flash.index()('somemodule')`. Is there any real reason for that much indirection? It would seem much more "normal" to do `node.flash.index('somemodule')`.

Also, I'd much prefer to have the index itself returned as a table (`local t = node.flash.index()`) so it can be read programatically, compared to just printed out (what `node.flash.index()()` currently does).

Am I missing something here that would prevent this simplification?

Reply:
The previous fact that Lua ran from RAM and C ran from flash lead to the practice of "when in doubt, code it in C", even when a direct Lua approach is simpler and more transparent therefore easier to modify for the typical Lua developer.  In this case it would be easier for the Lua application programmer to say `lfs.someroutine(args)` to invoke `someroutine` in the **LFS**. Yes, we could hard core this in C but this is only a couple of lines of Lua which would be in your unit routine in the LFS, so the _only_ cludgy reference would be the first `node.flash.index()'init'()` call and if we allowed the user to define `LUA_INIT_STRING` in the config then with
```Lua
#define LUA_INIT_STRING "pcall(function()node.flash.index()'init'()end)"
```
then they wouldn't even need to do that.

Reply:
With the switch to LFS, I would also like to see a switch to the core dogma:

-   _Only use C if you can't do it in Lua or there are string performance reasons why it must be done in C._

Just think:  change a line of C and you have to reflash the entire firmware.  Change one line of Lua and you (or your provisioning subsystem) only needs to replace one SPIFFS file.  A mega hassle vs. a few seconds.

For example we've got a C ds18b20 module which is slow and far too limited functionality.  My Lua version is both richer and a lot faster.  So why use C?  The old answer was because C ran from flash so took up less RAM footprint.  This argument no longer applies.

As to this LFS init routine, if you do a `-p`  option on the `luac.cross` then you will see that this doesn't use a table which some C routine parses, the code is a directly emitted Lua function.

And as I have said elsewhere is really that hard to include something like the following in the LFS:init module?
```Lua
do
  local ndx = node.flash.index 
  local lfs_t = { __index = function(_, name)
      local fn = ndx(name)
      if type(fn) == 'function' then return fn end -- or return nil implied
    end}
  getfenv().lfs = setmetatable(lfs_t,lfs_t)
end
``` 
The RAM footprint is a table with a single entry in it and a LClosure with a single upval.  The LFS footprint is a code fragment of 16 instructions and a function of 10 instructions.

Reply:
You earlier argued that both mine and your uses were "advanced" (probably rightly so), so I think I'll level that same comment back at you here :)  Meta tables aren't exactly the most obvious things when one starts out with Lua.

I personally was looking at `node.flash.index` as a drop-in replacement for `dofile` (or even `require`). It's something existing developers are familiar with both concept wise and syntax wise, and it matches how the files/modules get interned into the LFS. Where before you would've `mymod = dofile('mymod.lc')` you could (assuming suggested change) `mymod = node.flash.index('mymod')`. To me that would be a nice path into the LFS.

To address the dogma shift, yes, I agree in principle, but we'd need to provide an easy way for users to build the `flash.img`. Right now Marcel's cloudbuilder is saving everyone from needing their own build environment, which is fabulous from a usability perspective. It does make it hard to start using LFS with this though. Say @marcelstoer, do you think it would be feasible to also have an LFS builder service? A user uploads their `.lua` files and we spit out a either a PIC `flash.img` to be `node.flash.reload()`ed or an `lfs.bin` to be `esptool`'d straight onto the device? If I had time and somewhere to host it I could probably put something together (insert bemoaning of lack of time here).

Reply:
This is my `init.lua`

```
if file.rename('flash.img', 'loaded.img') then
  node.flash.reload('loaded.img')
end
node.flash.index()('init')()
```

Then my `init.lua` in the ROM starts with

```
local index = node.flash.index and node.flash.index()
local function loader_flash(module)
  local r = index(module)
  return type(r) == 'function' and r -- or nil otherwise
end
if index then package.loaders[3] = loader_flash end
```

Reply:
@pjsq Maybe you can explain what I'm missing here? Why do we need to do
```
local index = node.flash.index and node.flash.index()
```
rather than
```
local index = node.flash.index
```
What feature is gained by requiring that function call? Why can't that be hidden inside the `node.flash.index` function itself?

Reply:
@jmattsson the and trick is simple because 
-  if the LFS isn't compiled in then `node.flash` returns `nil` and so evaluating `node.flash.index` will throw an exception, causing a panic.
-  if the LFS is compiled in, but the LFS isn't loaded, then `node.flash.index` returns `nil` and so evaluating `node.flash.index()` will throw an exception, causing a panic.  In the case of running _from LFS_, the `node.flash.index and ` is redundant because this code can assume that LFS is both compiled and loaded, and this makes life a lot simpler.  

My preferred `init.lua`approach is:
```Lua
pcall(function() local f=node.flash; return f.index and f.index()('init')() or f.reload('.img') end)
```
which is guaranteed not to panic. and I do the update reload using my previsioning system. 

I chose the Lua route because (a) it is simple; and (b) it leaves it to the individual developer / project to implement their own Lua encapsulation that's right for them.  This got the patch out there to evaluate.  I don't think that using meta tables is that much of an issue as we can include a default in Lua examples and refer to it in the API reference. Novice Lua programmers can use it as-is and more advanced ones can use it as a starting point to modify.

I will give an analogy: Enduser_setup is written in C, by I regularly see Lua developers post issues "I don't know how to write C modules, but how do I modify it to do XYZ?"  If it was in Lua, then they could just tweak the code.

----
This all being said, changing the API is pretty straight forward, so long as we first agree what we want to change it to.  Let's pick up this general point on #2292